src/ralph/models/xapi/base/common.py

```diff
@@ -1,52 +1,54 @@
-from pydantic import StrictStr, validate_email
+from pydantic import Field, RootModel, model_validator
 from rfc3987 import parse
 
+from ralph.conf import NonEmptyStrictStr
 
-class IRI(str):
+
+class IRI(RootModel[Union["IRI", str]]):
     """Pydantic custom data type validating RFC 3987 IRIs."""
 
-    @classmethod
-    def __get_validators__(cls) -> Generator:  # noqa: D105
-        def validate(iri: str) -> Type["IRI"]:
-            """Check whether the provided IRI is a valid RFC 3987 IRI."""
-            parse(iri, rule="IRI")
-            return cls(iri)
+    def __hash__(self):  # noqa: D105
+        return hash(str(self.root))
 
-        yield validate
+    def __str__(self):  # noqa: D105
+        return str(self.root)
 
+    @model_validator(mode="before")
+    @classmethod
+    def validate_iri(cls, iri):
+        """Check whether the provided IRI is a valid RFC 3987 IRI."""
+        parse(str(iri), rule="IRI")
+        return str(iri)
 
-class LanguageTag(str):
-    """Pydantic custom data type validating RFC 5646 Language tags."""
 
-    @classmethod
-    def __get_validators__(cls) -> Generator:  # noqa: D105
-        def validate(tag: str) -> Type["LanguageTag"]:
-            """Check whether the provided tag is a valid RFC 5646 Language tag."""
-            if not tag_is_valid(tag):
-                raise TypeError("Invalid RFC 5646 Language tag")
-            return cls(tag)
+class LanguageTag(RootModel[Union[str, "LanguageTag"]]):  # TODO: verify this
+    """Pydantic custom data type validating RFC 5646 Language tags."""
 
-        yield validate
+    def __hash__(self):  # noqa: D105
+        return hash(str(self.root))
 
+    def __str__(self):  # noqa: D105
+        return str(self.root)
 
-LanguageMap = Dict[LanguageTag, StrictStr]
+    @model_validator(mode="before")
+    @classmethod
+    def validate_language_tag(cls, tag):
+        """Check whether the provided tag is a valid RFC 5646 Language tag."""
+        if not tag_is_valid(str(tag)):
+            raise TypeError("Invalid RFC 5646 Language tag")
+        return str(tag)
 
 
-class MailtoEmail(str):
-    """Pydantic custom data type validating `mailto:email` format."""
+LanguageMap = Dict[LanguageTag, NonEmptyStrictStr]  # TODO: change back to strictstr
 
-    @classmethod
-    def __get_validators__(cls) -> Generator:  # noqa: D105
-        def validate(mailto: str) -> Type["MailtoEmail"]:
-            """Check whether the provided value follows the `mailto:email` format."""
-            if not mailto.startswith("mailto:"):
-                raise TypeError("Invalid `mailto:email` value")
-            valid = validate_email(mailto[7:])
-            return cls(f"mailto:{valid[1]}")
-
-        yield validate
+email_pattern = (
+    r"(^mailto:[-!#-'*+/-9=?A-Z^-~]+(\.[-!#-'*+/-9=?A-Z^-~]+)*|\"([]!#-[^-~ \t]|"
+    r"(\\[\t -~]))+\")@([-!#-'*+/-9=?A-Z^-~]+(\.[-!#-'*+/-9=?A-Z^-~]+)*|\[[\t -Z^-~]*])"
+)
+MailtoEmail = Annotated[str, Field(pattern=email_pattern)]
```

Comment:
```python
TypeAdapter(MailtoEmail).validate_python('" "@!')
```
This example seems to not raise an exception.
What about using pydantic's `email-validator`?